cli/autoconfig/autoconfig.go

```diff
@@ -57,41 +60,77 @@ func remoteBazelArgs(besBackend, remoteExecutor string) []string {
+	netIface, err := networking.DefaultInterface(ctx)
+	if err != nil {
+		return nil, err
+	}
+	addrs, err := netIface.Addrs()
+	if err != nil {
+		return nil, err
+	}
+	for _, a := range addrs {
+		if v, ok := a.(*net.IPNet); ok {
+			return v.IP, nil
+		}
 	}
+	return nil, fmt.Errorf("could not determine host's default IP")
+}
 
-	wsFilePath, err := findWorkspaceFile()
+// TODO(vadim): reduce # of return values
+func Configure(ctx context.Context, bazelFlags *commandline.BazelFlags, filteredOSArgs []string) (*commandline.BazelFlags, *BazelOpts, []string, error) {
+	grpcEndpoint := ""
+	webURL := ""
+	switch *env {
+	case "prod":
+		grpcEndpoint = "grpcs://remote.buildbuddy.io"
+		webURL = "https://app.buildbuddy.io"
+	case "dev":
+		grpcEndpoint = "grpcs://remote.buildbuddy.dev"
+		webURL = "https://app.buildbuddy.dev"
+	case "local":
+		// we need to use an IP that's reachable from within the firecracker VMs
+		ip, err := defaultIP(ctx)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		grpcEndpoint = fmt.Sprintf("grpc://%s:1985", ip)
+		webURL = "http://localhost:8080"
+	default:
+		return nil, nil, nil, fmt.Errorf("unknown environment %q", *env)
+	}
+
+	wsFilePath, err := bazel.FindWorkspaceFile(".")
 	if err != nil {
 		return nil, nil, nil, err
 	}
 
 	if *events && bazelFlags.BESBackend == "" {
-		bazelFlags.BESBackend = fmt.Sprintf("grpcs://remote.%s", serviceDomain)
-		filteredOSArgs = append(filteredOSArgs, fmt.Sprintf("--bes_results_url=https://app.%s/invocation/", serviceDomain))
+		bazelFlags.BESBackend = grpcEndpoint
+		filteredOSArgs = append(filteredOSArgs, fmt.Sprintf("--bes_results_url=%s/invocation/", webURL))
 	}
 
 	if (*remote == remoteExec || *remote == remoteBazel) && bazelFlags.RemoteExecutor == "" {
-		remoteExecTarget := fmt.Sprintf("grpcs://remote.%s", serviceDomain)
 		if *remote == remoteExec {
 			filteredOSArgs = append(filteredOSArgs, bbToolchainArgs()...)
 			addToolchainsToWorkspaceIfNotPresent(wsFilePath)
-			bazelFlags.RemoteExecutor = remoteExecTarget
+			bazelFlags.RemoteExecutor = grpcEndpoint
+			filteredOSArgs = append(filteredOSArgs, "--remote_download_minimal")
 		} else {
-			filteredOSArgs = append(filteredOSArgs, remoteBazelArgs(bazelFlags.BESBackend, remoteExecTarget)...)
+			filteredOSArgs = append(filteredOSArgs, remoteBazelArgs(bazelFlags.BESBackend, grpcEndpoint)...)
+			// --remote_download_minimal is supposed to be enough for `bazel run` to work, but it's not the case in
+			// practice. Use --remote_download_toplevel to make sure the top level binary is available for running.
+			filteredOSArgs = append(filteredOSArgs, "--remote_download_toplevel")
 		}
-		filteredOSArgs = append(filteredOSArgs, "--remote_download_minimal")
 		filteredOSArgs = append(filteredOSArgs, "--jobs=200")
 	}
 
 	if (*cache || *remote == remoteBazel) && bazelFlags.RemoteCache == "" {
-		bazelFlags.RemoteCache = fmt.Sprintf("grpcs://remote.%s", serviceDomain)
+		bazelFlags.RemoteCache = grpcEndpoint
 	}
```

Comment:
<not this cl> but it seems like there's a lot going on here

Reply:
I agree. (this whole func can use some love, i will take a look after i'm done with the run changes)